pkg/controller/logstorage/users/user_controller.go

```diff
@@ -0,0 +1,254 @@
+	basicCreds := corev1.Secret{}
+	credentialSecrets := []client.Object{}
+	key = types.NamespacedName{Name: render.ElasticsearchLinseedUserSecret, Namespace: helper.TruthNamespace()}
+	if err := r.client.Get(ctx, key, &basicCreds); err != nil && !errors.IsNotFound(err) {
+		r.status.SetDegraded(operatorv1.ResourceReadError, fmt.Sprintf("Error getting Secret %s", key), err, reqLogger)
+		return reconcile.Result{}, err
+	} else if errors.IsNotFound(err) {
+		// Create the secret to provision into the cluster.
+		pw, err := randomPassword(16)
+		if err != nil {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "Error generating Linseed password", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+		basicCreds = corev1.Secret{}
+		basicCreds.Name = render.ElasticsearchLinseedUserSecret
+		basicCreds.Namespace = helper.TruthNamespace()
+		basicCreds.StringData = map[string]string{"username": linseedUser.Username, "password": pw}
+
+		// Make sure we install the generated credentials into the truth namespace.
+		credentialSecrets = append(credentialSecrets, &basicCreds)
+	}
+	if helper.TruthNamespace() != helper.InstallNamespace() {
+		// Copy the credentials into the install namespace.
+		credentialSecrets = append(credentialSecrets, secret.CopyToNamespace(helper.InstallNamespace(), &basicCreds)[0])
+	}
+	credentialComponent := render.NewPassthrough(credentialSecrets...)
+
+	// In standard installs, the LogStorage owns the secret. For multi-tenant, it's owned by the tenant.
+	var hdler utils.ComponentHandler
+	if r.multiTenant {
+		hdler = utils.NewComponentHandler(reqLogger, r.client, r.scheme, tenant)
+	} else {
+		hdler = utils.NewComponentHandler(reqLogger, r.client, r.scheme, logStorage)
+	}
+	if err := hdler.CreateOrUpdateOrDelete(ctx, credentialComponent, r.status); err != nil {
+		r.status.SetDegraded(operatorv1.ResourceUpdateError, "Error creating / updating Linseed user secret", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Now that the secret has been created, also provision the user in ES.
+	if err := r.createLinseedLogin(ctx, tenantID, &basicCreds, reqLogger); err != nil {
+		r.status.SetDegraded(operatorv1.ResourceUpdateError, "Failed to create Linseed user in ES", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	r.status.ReadyToMonitor()
+	r.status.ClearDegraded()
+	return reconcile.Result{}, nil
+}
+
+func (r *UserController) createLinseedLogin(ctx context.Context, tenantID string, secret *corev1.Secret, reqLogger logr.Logger) error {
+	// ES should be in ready phase when execution reaches here, apply ILM polices
+	esClient, err := utils.NewElasticClient(r.client, ctx, relasticsearch.ElasticEndpoint())
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceCreateError, "Failed to connect to Elasticsearch - failed to create the Elasticsearch client", err, reqLogger)
+		return err
+	}
+
+	// Determine the password from the secret.
+	password := secret.StringData["password"]
```

Comment:
```
	// stringData allows specifying non-binary secret data in string form.
	// It is provided as a write-only input field for convenience.
	// All keys and values are merged into the data field on write, overwriting any existing values.
	// The stringData field is never output when reading from the API.
```

Reply:
Yep. The code above may pass either a queried secret (in which case the data is in `Data` or a freshly created one, in which case the data is in `StringData`

Reply:
I think we best refrain from using this "convenience" function if we want to read the contents before writing.

Reply:
We're not reading the contents before writing. We do this:

1. Get the secret if it exists.
2. If it doesn't exist, make a new one.
3. Write it to the API.
4. Assuming it was successful, create the user in ES

This function is step 4 - we're just avoiding the need to wait for our watches to sync / re-query the secret, since we just wrote it, and so we know exactly what the values are.

If we reach step 4 via step 2, then `StringData` will have the fields present.

If we reach step 4 via step 1, then `Data` will have them present.

Reply:
My concern remains that the use of StringData has made the code more complicated.

Reply:
It is marginally more complicated in this particular function - it's an extra three lines of code in a helper function - to avoid complexity elsewhere. Namely, we would need additional logic in the calling code to handle this instead.

Reply:
Other options:

1. Just error out the first time this is called, logging an error and degrading status, wait for the retry to fix it.

2. Update the calling code logic to short-circuit if this is the first time this has been called. For example:

```go

creatingSecret  := false
if err := r.client.Get(ctx, key, &basicCreds); err != nil && !errors.IsNotFound(err) {
  r.status.SetDegraded(operatorv1.ResourceReadError, fmt.Sprintf("Error getting Secret %s", key), err, reqLogger)
  return reconcile.Result{}, err
} else if errors.IsNotFound(err) {
  creatingSecret = true
}
...

if err := hdler.CreateOrUpdateOrDelete(ctx, credentialComponent, r.status); err != nil {
  r.status.SetDegraded(operatorv1.ResourceUpdateError, "Error creating / updating Linseed user secret", err, reqLogger)
  return reconcile.Result{}, err
}

if creatingSecret {
  // We just created the secret. Return early, and we'll provision it to ES the next time reconcile is called.
  return reconcile.Result{Requeue: 5 * time.Second}, nil
}

// Create the user in ES
...

```

3. Pass the username and password to this function explicitly instead of the secret

```go

username := linseedUser.Username
password := "" // To be filled in below.
if err := r.client.Get(ctx, key, &basicCreds); err != nil && !errors.IsNotFound(err) {
  r.status.SetDegraded(operatorv1.ResourceReadError, fmt.Sprintf("Error getting Secret %s", key), err, reqLogger)
  return reconcile.Result{}, err
} else if errors.IsNotFound(err) {
  // Create the secret
  password = RandomPassword(16)
} else {
  // The secret already exists. Get the password from it.
  password = string(basicCreds.Data["password"])
}
...

// Create the user in ES
createLinseedLogin(username, password)
...

```

Of these, option 3 is the only one that I think could be considered less complicated than how it is currently implemented. I can make that switch, but I still very slightly prefer the way it is now because the complexity is limited to one small code block within a helper function.